rasa/core/featurizers/tracker_featurizers.py

```diff
@@ -528,100 +710,439 @@ def slice_state_history(
-                tracker, domain, omit_unset_slots=omit_unset_slots
-            )
 
-            states_length_for_action = 0
-            entity_data = {}
-            for event in tracker.applied_events():
-                if isinstance(event, UserUttered):
-                    entity_data = self._entity_data(event)
+            for states, label, entities in self._extract_examples(
+                tracker,
+                domain,
+                omit_unset_slots=omit_unset_slots,
+                ignore_action_unlikely_intent=ignore_action_unlikely_intent,
+            ):
 
-                if not isinstance(event, ActionExecuted):
+                if self.remove_duplicates:
+                    hashed = self._hash_example(tracker, states, label)
+                    if hashed in hashed_examples:
+                        continue
+                    hashed_examples.add(hashed)
+
+                example_states.append(states)
+                example_labels.append(label)
+                example_entities.append(entities)
+
+                pbar.set_postfix({f"# {self.LABEL_NAME}": f"{len(example_labels):d}"})
+
+        self._remove_user_text_if_intent(example_states)
+
+        logger.debug(f"Created {len(example_states)} {self.LABEL_NAME} examples.")
+
+        return example_states, example_labels, example_entities
+
+    def _extract_examples(
+        self,
+        tracker: DialogueStateTracker,
+        domain: Domain,
+        omit_unset_slots: bool = False,
+        ignore_action_unlikely_intent: bool = False,
+    ) -> Iterator[Tuple[List[State], List[Text], List[Dict[Text, Any]]]]:
+        """Create an iterator over training examples from a tracker.
+
+        Create an iterator over the training examples that will be created
+        from the provided tracker.
+
+        Args:
+            trackers: The tracker from which to extract training examples.
+            domain: The domain of the training data.
+            omit_unset_slots: If `True` do not include the initial values of slots.
+            ignore_action_unlikely_intent: Remove action_unlikely_intent from
+                training states.
+
+
+
+        Returns:
+            An iterator over tuples of list of states, list of labels and list of
+                entity data.
+        """
+        tracker_states = self._create_states(
+            tracker, domain, omit_unset_slots=omit_unset_slots
+        )
+
+        if ignore_action_unlikely_intent:
+            tracker_states = self._remove_action_unlikely_intent(tracker_states)
+
+        label_index = 0
+        entity_data = {}
+        for event in tracker.applied_events():
+            if isinstance(event, UserUttered):
+                entity_data = self._entity_data(event)
+
+            elif isinstance(event, ActionExecuted):
+
+                if (
+                    ignore_action_unlikely_intent
+                    and event.action_name == ACTION_UNLIKELY_INTENT_NAME
+                ):
```

Comment:
Okay, I can see why you need the `if` construct here. You want to do the cleanup before the states are sliced. In that case I would recommend moving the call to `slice_state_history` outside of this function and move it inside the `for` loop (line 781) of `training_states_labels_and_entities`. So, `_extract_examples` can operate as if it knows nothing about slicing but then `training_states_label_and_entities` applies the slicing as needed.
 What do you think?